circuits/src/zk_circuits/valid_wallet_update.rs

```diff
@@ -0,0 +1,1242 @@
+            // another below
+            let equals_external_transfer_mint = OrGate::or(
+                EqGadget::eq(old_balance.mint.clone(), external_transfer.mint.into(), cs),
+                EqGadget::eq(new_balance.mint.clone(), external_transfer.mint.into(), cs),
+                cs,
+            );
+
+            let (_, _, external_transfer_term) = cs.multiply(
+                equals_external_transfer_mint.clone(),
+                external_transfer_term.clone(),
+            );
+
+            external_transfer_applied += equals_external_transfer_mint;
+            expected_balance += external_transfer_term;
+
+            // Constrain the updated balance to be non-negative and correctly updated
+            cs.constrain(new_balance.amount.clone() - expected_balance);
+            GreaterThanEqZeroGadget::<64 /* bitwidth */>::constrain_greater_than_zero(
+                new_balance.amount.clone(),
+                cs,
+            );
+
+            // Constrain the mints to be set correctly. A valid mint is one of:
+            //  1. The same mint as in the old wallet, if the balance previously existed and was not zero'd
+            //  2. The mint of the transfer if the balance was previously zero; balances are constrained
+            //     to be unique elsewhere in the circuit
+            //  3. The zero mint, if the balance is zero
+            let mints_equal = EqGadget::eq(old_balance.mint.clone(), new_balance.mint.clone(), cs);
+            let equals_transfer_mint =
+                EqGadget::eq(new_balance.mint.clone(), external_transfer.mint.into(), cs);
+            let mint_is_zero = EqZeroGadget::eq_zero(new_balance.mint.clone(), cs);
+            let new_balance_zero = EqZeroGadget::eq_zero(new_balance.amount.clone(), cs);
+            let prev_balance_zero = EqZeroGadget::eq_zero(old_balance.amount.clone(), cs);
+
+            // Condition 1 -- same mint as old wallet, balance not zero'd
+            let valid_mint1 = AndGate::multi_and(
+                &[
+                    mints_equal.into(),
+                    NotGate::not(prev_balance_zero, cs),
+                    NotGate::not(new_balance_zero, cs),
+                ],
+                cs,
+            );
+
+            // Condition 2 -- new mint added to wallet
+            let valid_mint2 = AndGate::multi_and(
+                &[
+                    equals_transfer_mint.into(),
+                    prev_balance_zero.into(),
+                    NotGate::not(new_balance_zero, cs),
+                ],
+                cs,
+            );
+
+            // Condition 3 -- withdrawal of entire balance, mint is now zero
+            let valid_mint3 = AndGate::multi_and(&[mint_is_zero, new_balance_zero], cs);
+
+            // Constrain one of the three mint conditions to hold
+            let valid_mint = OrGate::multi_or(&[valid_mint1, valid_mint2, valid_mint3], cs);
```

Comment:
should this be an `XOR`?
i get that logically it's impossible for any two of these clauses to be true at the same time, but an adversarial prover could try to set `valid_mint_i = valid_mint_j = 1` since they're new wires in the constraint system, no?

Reply:
Talked in person, copying here for transparency -- the inputs to the `OR` gate are implicitly mutually exclusive, so `OR` and `XOR` are equivalent. `OR` is more efficient